service/udp.go

```diff
@@ -135,7 +135,7 @@ func (s *udpService) Serve(clientConn net.PacketConn) error {
 			}()
 
 			// Attempt to read an upstream packet.
-			clientProxyBytes, clientAddr, err := clientConn.ReadFrom(cipherBuf)
+			clientProxyBytes, clientAddr, proxyIP, err := onet.ReadFromWithDst(clientConn, cipherBuf)
```

Comment:
This change is intrusive and adds another thing to keep track of. I think we can do better by composing behavior.
Instead of returning the `proxyIP`, can you return a "bound" `clientConn` that holds the `proxyIP` and `clientAddr` internally and uses that for all the subsequent writes and reads? .
This way we don't need to pass `proxyIP` and `clientAddr` around.

You will still need to expose those fields for `makeNATKey` though.

Reply:
A full version of this would require moving the NAT code from `service/` to `net/`, leaving the UDP code in `service/` looking much like the current TCP code.  I see the appeal of this, but after attempting to implement it I think it's not worthwhile.  In particular, the multiplexing of blocking reads from multiple underlying sockets generates a lot of complexity, especially if we try to avoid allocating a fresh buffer for every packet (as in the current design).

After considering various options, I think the current design provides the best balance of simplicity and performance.  I've reorganized the functionality to present a cleaner interface, but I have not attempted to provide any "virtual Conn" wrapping.